test/e2e/authorization_utils.go

```diff
@@ -0,0 +1,278 @@
+type testCase struct {
+	data   testcaseData
+	output testcaseOutput
+}
+
+// testcaseData is a struct that makes it user-friendly to write testcases
+// more logically. This will be used to generate individual SubjectAccessReview
+// objects in order to test the authorization rules.
+type testcaseData struct {
+	namespaces       []string
+	names            []string
+	verbs            []string
+	apiGroups        []string
+	resources        []string
+	subresources     []string
+	nonResourceVerbs []string
+	nonResourcePaths []string
+	users            []string
+
+	// this is double slice since we need to check individually for
+	// each group of users. A single slice would mean that the same user
+	// is part of all the groups.
+	groups [][]string
+}
+
+// testcaseOutput is a struct that represents the expected result of a testcase.
+// This is based on the SubjectAccessReviewStatus type. It provides simplicity in testcase
+// writing since one testcase can have multiple SubjectAccessReview objects
+// and we need to determine the "final" result and expose that as the testcase output.
+type testcaseOutput struct {
+	// the final result based on results of individual SubjectAccessReview objects
+	allowed, denied bool
+	// the set of reasons from individual SubjectAccessReview objects
+	reason []string
+}
+
+func (t *testCase) run(ctx context.Context, cs kubernetes.Interface) error {
+	// Generate the list of SubjectAccessReview objects based on the testcase data
+	sars := t.generateSubjectAccessReviews()
+
+	// Create the SubjectAccessReview objects in the cluster
+	createdSars, err := createSubjectAccessReviews(ctx, cs, sars)
+	if err != nil {
+		return err
+	}
+
+	// Evaluate the output based on the created SubjectAccessReview objects
+	// and set the final result in the testcase output
+	t.evaluateOutput(createdSars)
+
+	return nil
+}
+
+// accessReviewGenerator generates a list of SubjectAccessReview objects based on the
+// testcase data provided.
+func (t *testCase) generateSubjectAccessReviews() []authv1.SubjectAccessReview {
+	// Initialize the list of SubjectAccessReview objects
+	sars := make([]authv1.SubjectAccessReview, 0)
+
+	// expand the testcase data to generate a list of ResourceAttributes
+	resourceAttributes := t.expandResourceAttributes()
+
+	// expand the testcase data to generate a list of NonResourceAttributes
+	// nonResourceAttributes := t.expandNonResourceAttributes()
+
+	// expand the testcase data to generate a list of SubjectAccessReview objects
+	// based on the ResourceAttributes and NonResourceAttributes
+	for _, ra := range resourceAttributes {
+		for _, user := range t.data.users {
+			for _, group := range t.data.groups {
+				sar := authv1.SubjectAccessReview{
+					Spec: authv1.SubjectAccessReviewSpec{
+						ResourceAttributes: &ra,
+						User:               user,
+						Groups:             group,
+					},
+				}
+				sars = append(sars, sar)
+			}
+		}
+	}
+	return sars
+}
+
```

Comment:
This is a temporary fix to get the tests running, will be refactored into expansions.

Reply:
This is now done.